custom/ewsghana/reports/email_reports.py

```diff
@@ -0,0 +1,222 @@
+                monthly_consumption=0
+            ) for product in products
+        }
+
+        for location in locations:
+            stock_states = StockState.objects.filter(
+                case_id=location.supply_point_id,
+                section_id=STOCK_SECTION_TYPE,
+                sql_product__in=products
+            ).order_by('-last_modified_date')
+
+            for state in stock_states:
+                p_name = state.sql_product.name
+                row_data[p_name]['total_fac'] += 1
+                if state.stock_on_hand:
+                    row_data[p_name]['total_stock'] += state.stock_on_hand
+                else:
+                    row_data[p_name]['fac_with_stockout'] += 1
+                if state.get_monthly_consumption():
+                    row_data[p_name]['fac_with_consumption'] += 1
+                    row_data[p_name]['monthly_consumption'] += state.get_monthly_consumption()
+
+        rows = []
+        for k, v in row_data.iteritems():
+            months_of_stock = float(v['total_stock']) / float(v['monthly_consumption'] or 1)
+            rows.append([k, percent(v['fac_with_stockout'], v['total_fac']),
+                        v['total_fac'], v['total_stock'], percent(v['fac_with_consumption'], v['total_fac']),
+                        v['monthly_consumption'], "%.1f" % months_of_stock, stock_status(months_of_stock)])
+        return rows
+
+
+class StockSummaryReportData(EmailReportData):
+    title = 'Weekly Stock Summary Report - '
+    slug = 'stock_summary'
+
+    def get_locations(self, loc_id, domain):
+        self.title += SQLLocation.objects.get(location_id=loc_id).name
+        return get_stock_summary_locations(loc_id, domain)
+
+
+class CMSRMSReportData(EmailReportData):
+    title = 'Weekly Stock Summary Report - CMS and RMS'
+    slug = 'stock_summary'
+
+    def get_locations(self, loc_id, domain):
+        return get_rms_cms_summary_locations(domain)
+
+
+class EmailReportingData(EWSData):
+    title = "Reporting"
+    slug = "reporting"
+    show_table = True
+
+    def get_locations(self, loc_id, domain):
+        return []
+
+    @property
+    def headers(self):
+        return DataTablesHeader(*[
+            DataTablesColumn(_('% of facilities are reporting')),
+            DataTablesColumn(_('% of submitted reports are complete')),
+        ])
+
+    @property
+    def rows(self):
+        def percent(x, y):
+            return "%d%% (%d/%d)" % (x * 100 / (y or 1), x, y)
+
+        locations = self.get_locations(self.config['location_id'], self.config['domain'])
+        reported = StockTransaction.objects.filter(
+            case_id__in=locations, report__date__range=[self.config['startdate'],
+                                                        self.config['enddate']]).distinct('case_id').count()
+        completed = 0
+        for loc in locations:
+                products_count = len(SQLLocation.objects.get(supply_point_id=loc).products)
```

Comment:
I think you could optimize this by pushing the count to the database by referencing `._products` if this project uses location-specific products, otherwise the domain setting.